### ISSUE TITLE
alfa-python make error

### ISSUE THREAD

**Issue by Angi2412** [castacks/alfa-dataset-tools#2]
Hello,

I tried to make the alfa-python package in MacOS as well as in Windows.
In Windows, it says that `boost/python.hpp` is not present.
And in MacOS, I got the following output:

```
Last login: Tue Sep 15 10:27:19 on ttys001
angi@Angis-MacBook-Pro ~ % cd Uni 
angi@Angis-MacBook-Pro Uni % git clone https://github.com/castacks/alfa-dataset-tools.git
Cloning into 'alfa-dataset-tools'...
remote: Enumerating objects: 20, done.
remote: Counting objects: 100% (20/20), done.
remote: Compressing objects: 100% (17/17), done.
remote: Total 471 (delta 4), reused 17 (delta 3), pack-reused 451
Receiving objects: 100% (471/471), 6.77 MiB | 7.80 MiB/s, done.
Resolving deltas: 100% (232/232), done.
angi@Angis-MacBook-Pro Uni % cd alfa-dataset-tools/alfa-python 
angi@Angis-MacBook-Pro alfa-python % mkdir build
angi@Angis-MacBook-Pro alfa-python % cd build
angi@Angis-MacBook-Pro build % cmake ..
-- The C compiler identification is AppleClang 11.0.3.11030032
-- The CXX compiler identification is AppleClang 11.0.3.11030032
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PythonInterp: /usr/bin/python (found version "2.7.16") 
-- Found PythonLibs: /usr/lib/libpython2.7.dylib (found version "2.7.16") 
-- Using Python2
-- Configuring done
CMake Warning (dev):
  Policy CMP0042 is not set: MACOSX_RPATH is enabled by default.  Run "cmake
  --help-policy CMP0042" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  MACOSX_RPATH is not specified for the following targets:

   alfa_python

This warning is for project developers.  Use -Wno-dev to suppress it.

-- Generating done
-- Build files have been written to: /Users/angi/Uni/alfa-dataset-tools/alfa-python/build
angi@Angis-MacBook-Pro build % make
Scanning dependencies of target Boost
[ 10%] Creating directories for 'Boost'
[ 20%] Performing download step (download, verify and extract) for 'Boost'
-- Downloading...
   dst='/Users/angi/Uni/alfa-dataset-tools/alfa-python/build/Boost/src/boost_1_63_0.tar.bz2'
   timeout='none'
-- Using src='http://sourceforge.net/projects/boost/files/boost/1.63.0/boost_1_63_0.tar.bz2/download'
-- [download 100% complete]
-- [download 0% complete]
-- [download 1% complete]
-- [download 2% complete]
-- [download 3% complete]
-- [download 4% complete]
-- [download 5% complete]
-- [download 6% complete]
-- [download 7% complete]
-- [download 8% complete]
-- [download 9% complete]
-- [download 10% complete]
-- [download 11% complete]
-- [download 12% complete]
-- [download 13% complete]
-- [download 14% complete]
-- [download 15% complete]
-- [download 16% complete]
-- [download 17% complete]
-- [download 18% complete]
-- [download 19% complete]
-- [download 20% complete]
-- [download 21% complete]
-- [download 22% complete]
-- [download 23% complete]
-- [download 24% complete]
-- [download 25% complete]
-- [download 26% complete]
-- [download 27% complete]
-- [download 28% complete]
-- [download 29% complete]
-- [download 30% complete]
-- [download 31% complete]
-- [download 32% complete]
-- [download 33% complete]
-- [download 34% complete]
-- [download 35% complete]
-- [download 36% complete]
-- [download 37% complete]
-- [download 38% complete]
-- [download 39% complete]
-- [download 40% complete]
-- [download 41% complete]
-- [download 42% complete]
-- [download 43% complete]
-- [download 44% complete]
-- [download 45% complete]
-- [download 46% complete]
-- [download 47% complete]
-- [download 48% complete]
-- [download 49% complete]
-- [download 50% complete]
-- [download 51% complete]
-- [download 52% complete]
-- [download 53% complete]
-- [download 54% complete]
-- [download 55% complete]
-- [download 56% complete]
-- [download 57% complete]
-- [download 58% complete]
-- [download 59% complete]
-- [download 60% complete]
-- [download 61% complete]
-- [download 62% complete]
-- [download 63% complete]
-- [download 64% complete]
-- [download 65% complete]
-- [download 66% complete]
-- [download 67% complete]
-- [download 68% complete]
-- [download 69% complete]
-- [download 70% complete]
-- [download 71% complete]
-- [download 72% complete]
-- [download 73% complete]
-- [download 74% complete]
-- [download 75% complete]
-- [download 76% complete]
-- [download 77% complete]
-- [download 78% complete]
-- [download 79% complete]
-- [download 80% complete]
-- [download 81% complete]
-- [download 82% complete]
-- [download 83% complete]
-- [download 84% complete]
-- [download 85% complete]
-- [download 86% complete]
-- [download 87% complete]
-- [download 88% complete]
-- [download 89% complete]
-- [download 90% complete]
-- [download 91% complete]
-- [download 92% complete]
-- [download 93% complete]
-- [download 94% complete]
-- [download 95% complete]
-- [download 96% complete]
-- [download 97% complete]
-- [download 98% complete]
-- [download 99% complete]
-- [download 100% complete]
* Closing connection 2
* Closing connection 0
* Closing connection 1
-- verifying file...
       file='/Users/angi/Uni/alfa-dataset-tools/alfa-python/build/Boost/src/boost_1_63_0.tar.bz2'
-- Downloading... done
-- extracting...
     src='/Users/angi/Uni/alfa-dataset-tools/alfa-python/build/Boost/src/boost_1_63_0.tar.bz2'
     dst='/Users/angi/Uni/alfa-dataset-tools/alfa-python/build/Boost/src/Boost'
-- extracting... [tar xfz]
-- extracting... [analysis]
-- extracting... [rename]
-- extracting... [clean up]
-- extracting... done
[ 30%] No update step for 'Boost'
[ 40%] No patch step for 'Boost'
[ 50%] Performing configure step for 'Boost'
Building Boost.Build engine with toolset darwin... tools/build/src/engine/bin.macosxx86_64/b2
Detecting Python version... 2.7
Detecting Python root... /System/Library/Frameworks/Python.framework/Versions/2.7
Unicode/ICU support for Boost.Regex?... not found.
Generating Boost.Build configuration in project-config.jam...

Bootstrapping is done. To build, run:

    ./b2
    
To adjust configuration, edit 'project-config.jam'.
Further information:

   - Command line help:
     ./b2 --help
     
   - Getting started guide: 
     http://www.boost.org/more/getting_started/unix-variants.html
     
   - Boost.Build documentation:
     http://www.boost.org/build/doc/html/index.html

[ 60%] Performing build step for 'Boost'
Performing configuration checks

    - 32-bit                   : no
    - 64-bit                   : no
    - arm                      : no
    - mips1                    : no
    - power                    : no
    - sparc                    : no
    - x86                      : no
    - combined                 : no
    - symlinks supported       : yes

Component configuration:

    - atomic                   : not building
    - chrono                   : not building
    - container                : not building
    - context                  : not building
    - coroutine                : not building
    - coroutine2               : not building
    - date_time                : not building
    - exception                : not building
    - fiber                    : not building
    - filesystem               : not building
    - graph                    : not building
    - graph_parallel           : not building
    - iostreams                : not building
    - locale                   : not building
    - log                      : not building
    - math                     : not building
    - metaparse                : not building
    - mpi                      : not building
    - program_options          : not building
    - python                   : building
    - random                   : not building
    - regex                    : not building
    - serialization            : not building
    - signals                  : not building
    - system                   : not building
    - test                     : not building
    - thread                   : not building
    - timer                    : not building
    - type_erasure             : not building
    - wave                     : not building

clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
...skipped <pbin.v2/libs/python/build/darwin-11.0.3/release/link-static/threading-multi>libboost_python.a(clean) for lack of <pbin.v2/libs/python/build/darwin-11.0.3/release/link-static/threading-multi>numeric.o...
...skipped <pbin.v2/libs/python/build/darwin-11.0.3/release/link-static/threading-multi>libboost_python.a for lack of <pbin.v2/libs/python/build/darwin-11.0.3/release/link-static/threading-multi>numeric.o...
...skipped <p/Users/angi/Uni/alfa-dataset-tools/alfa-python/build/extern/lib>libboost_python.a for lack of <pbin.v2/libs/python/build/darwin-11.0.3/release/link-static/threading-multi>libboost_python.a...
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
clang: error: unknown argument: '-fcoalesce-templates'
...skipped <pbin.v2/libs/python/build/darwin-11.0.3/release/link-static/threading-multi>libboost_numpy.a(clean) for lack of <pbin.v2/libs/python/build/darwin-11.0.3/release/link-static/threading-multi>numpy/dtype.o...
...skipped <pbin.v2/libs/python/build/darwin-11.0.3/release/link-static/threading-multi>libboost_numpy.a for lack of <pbin.v2/libs/python/build/darwin-11.0.3/release/link-static/threading-multi>numpy/dtype.o...
...skipped <p/Users/angi/Uni/alfa-dataset-tools/alfa-python/build/extern/lib>libboost_numpy.a for lack of <pbin.v2/libs/python/build/darwin-11.0.3/release/link-static/threading-multi>libboost_numpy.a...
...failed updating 34 targets...
make[2]: *** [Boost/src/Boost-stamp/Boost-build] Error 1
make[1]: *** [CMakeFiles/Boost.dir/all] Error 2
make: *** [all] Error 2
angi@Angis-MacBook-Pro build % 
```
Maybe you can help me?
And Is there a plan for publishing a python package to pip in the (near) future?

**Comment by keipour** [comment]
Thank you for pointing out the issues. Looks like in Mac the issue was caused by the older boost library being confused between multiple compilers. I could solve the issue on my Mac by changing the version of the installed boost library. Please pull the latest repo update and check again. 
If you get an error with not finding `boost/python.hpp` you can try installing the `boost-python` module globally. If you still have issues after those steps please let us know. 

We will update the repo in 1-2 days with fixes to the other issues we found for alfa-python, including the issues in Windows and Ubuntu.

About the pip, we do have it in our to-do list, but I am not sure if that will happen soon.

**Comment by Angi2412** [comment]
So, I pulled the newest code from the repository and it solved the error on MacOS, thank you for the update.
Now I have the same error on both operating systems:

```
Scanning dependencies of target alfa_python
[ 90%] Building CXX object CMakeFiles/alfa_python.dir/alfa_python.cpp.o
/Users/angi/Uni/alfa-dataset-tools/alfa-python/alfa_python.cpp:24:10: fatal error: 
      'boost/python.hpp' file not found
#include <boost/python.hpp>
         ^~~~~~~~~~~~~~~~~~
1 error generated.
make[2]: *** [CMakeFiles/alfa_python.dir/alfa_python.cpp.o] Error 1
make[1]: *** [CMakeFiles/alfa_python.dir/all] Error 2
make: *** [all] Error 2
```

I already installed the `boost-python` (for Python 2 and 3) package via brew and also installed if from source and added the paths, but no change. I also used Python 2 as well as Python 3.

**Comment by Angi2412** [comment]
Any idea? :)

**Comment by keipour** [comment]
I pushed a fix for the issue and was able to compile it on my Linux system without any issues. If you still experience issues on Mac and Windows after the new fixes, please let us know so we can help you with solving the new problems. 
The current alfa-python version is basically a wrapper for the alfa-cpp module. We are most probably going to rewrite the whole module in a few weeks (the second half of October) directly in python to avoid the issues

**Comment by Angi2412** [comment]
Thank you for the support. The last commit solved the problem with the `boost/python.hpp` file.
But I had to edit two things in `BuildBoost.cmake` to make it work:
I had to add "27" to both linked packages otherwise I got an error (`ld: library not found for -lboost_python`) that these packages could not be found.
```Cmake
else()
  message(STATUS "Using Python2")
  set(Boost_LIBRARIES -lboost_python27 -lboost_numpy27)
endif()
```

Now after I got 100% progress from the make process, I have trouble importing the "alfa_python.so": It always says that there is no module name "alfa_python" even though it is in the same directory.
